File: 2020/python/day-02.py
# Read from input
with open("input.txt", "r") as input_file:
    _input = input_file.read().splitlines()

# Part 1
def part_1(_input):
    count = 0
    for i in _input:
        limits, char, password = i.split()

        lower_limit, upper_limit = [int(i) for i in limits.split("-")]
        char = char[0]

        if (password.count(char) in [i for i in range(lower_limit, upper_limit + 1)]):
            count += 1

    return count

# Part 2
def part_2(_input):
    count = 0
    for i in _input:
        indexes, char, password = i.split()

        index1, index2 = [(int(i) - 1) for i in indexes.split("-")]
        char = char[0]
        char_string = password[index1] + password[index2]

        if (char_string.count(char) == 1):
            count += 1

    return count

# Print answer
print(part_1(_input))
print(part_2(_input))
# [Finished in 434ms]

File: 2020/Go/day-02.go
package main
import (
    "bufio"
    "fmt"
    "os"
    "strconv"
    "strings"
)

// Part 1
func part1(input [][]string) (count int) {
    for _, value := range input {
        limit, char, password := strings.Split(value[0], "-"), string(value[1][0]), value[2]
        lower_limit, _ := strconv.Atoi(limit[0]); upper_limit, _ := strconv.Atoi(limit[1])
        char_count := strings.Count(password, char)

        if (char_count >= lower_limit && char_count <= upper_limit) {
            count++
        }
    }
    return
}

// Part 2
func part2(input [][]string) (count int) {
    for _, value := range input {
        indexes, char, password := strings.Split(value[0], "-"), string(value[1][0]), value[2]
        lower_index, _ := strconv.Atoi(indexes[0]); upper_index, _ := strconv.Atoi(indexes[1])
        char_string := string(password[lower_index - 1]) + string(password[upper_index - 1])

        if (strings.Count(char_string, char) == 1) {
            count++
        }
    }
    return
}


func main() {
    // Read data from input
    file, _ := os.Open("input.txt")
    defer file.Close()

    scanner := bufio.NewScanner(file)
    var input [][]string

    for scanner.Scan() {
        value := strings.Split(scanner.Text(), " ")
        input = append(input, value)
    }

    // Print answer
    fmt.Println(part1(input))
    fmt.Println(part2(input))
    // [Finished in 518.8µs]
}